   # Visual Data Analysis of Fraudulent Transactions

In [50]:
# initial imports
import pandas as pd
import calendar
import plotly.express as px
import hvplot.pandas
from sqlalchemy import create_engine
import datetime



In [51]:
# create a connection to the database
engine = create_engine("postgresql://postgres:Istay@10314@localhost:5432/fraud_detection")


In [52]:
# loading data from the database

def execute_query(query):

    transaction_df = pd.read_sql(sql=query, con=engine, index_col='date', parse_dates='date')

    return transaction_df

def fraud_transactions(df):
    start_time = datetime.time(7,0,0)
    end_time = datetime.time(9,0,0)
    return df.between_time(start_time, end_time).sort_values('amount', ascending=False)
    
    

In [58]:
# loading data for card holder 2 and 18 from the database
query = f'SELECT a.id, a.name, b.card, c.date, c.amount, e.name as "category" \
        FROM public.card_holder a, public.credit_card b, public.transaction c, public.merchant d, public.merchant_category e \
        WHERE a.id = b.id_card_holder AND b.card=c.card AND c.id_merchant=d.id AND d.id_merchant_category=e.id AND b.card=c.card'

transaction_df = execute_query(query)

suspect_df = fraud_transactions(transaction_df).head(100)
suspect_df




id               name                 card   amount  \
date                                                                       
2018-12-07 07:22:03   1     Robert Johnson  4761049645711555811  1894.00   
2018-03-05 08:26:08  16      Crystal Clark     5570600642865857  1617.00   
2018-03-06 07:18:09  25    Nancy Contreras        4319653513507  1334.00   
2018-01-22 08:07:03  16      Crystal Clark     5570600642865857  1131.00   
2018-09-26 08:48:40   1     Robert Johnson  4761049645711555811  1060.00   
2018-09-06 08:28:55   1     Robert Johnson  4761049645711555811  1017.00   
2018-03-26 07:41:59   9       Laurie Gibbs       30181963913340  1009.00   
2018-12-14 08:51:41  12        Megan Price         501879657465   748.00   
2018-04-01 07:17:21  25    Nancy Contreras        4319653513507   100.00   
2018-08-26 07:15:18  20      Kevin Spencer     4506405265172173    23.13   
2018-08-28 07:17:14  10  Matthew Gutierrez  4165305432349489280    20.71   
2018-10-07 08:16:54  20      Kevin Spencer  4586962917519654607    20.44   
2018-10-18 07:16:04  16      Crystal Clark     5500708021555307    19.86   
2018-03-03 08:42:02  24   Stephanie Dalton        4681896441519    19.50   
2018-06-04 08:30:28   4     Danielle Green         584226564303    19.49   
2018-03-26 08:08:27  16      Crystal Clark     5500708021555307    19.34   
2018-06-01 08:56:59  11     Brandon Pineda      180098539019105    19.33   
2018-12-23 07:39:47  10  Matthew Gutierrez  4962915017023706562    19.27   
2018-09-23 07:30:56  15        Kyle Tucker  4723783028106084756    19.02   
2018-02-21 08:21:10   5        Sara Cooper     4268491956169254    18.98   
2018-07-04 08:51:29  10  Matthew Gutierrez      213193946980303    18.93   
2018-08-02 07:44:24  19        Peter Mckay     5361779664174555    18.93   
2018-06-09 07:29:28  23         Mark Lewis     4188164051171486    18.82   
2018-06-02 07:45:48  23         Mark Lewis     4150721559116778    18.76   
2018-01-06 08:42:50  20      Kevin Spencer     3535651398328201    18.72   
2018-06-24 07:28:48  14        Gary Jacobs     5175947111814778    18.60   
2018-06-23 07:12:06   8      Michael Floyd     4834483169177062    18.56   
2018-08-14 08:38:49  18      Malik Carlson        4498002758300    18.54   
2018-10-07 07:08:08  24   Stephanie Dalton     3582198969197591    18.53   
2018-02-27 08:27:00   2      Shane Shaffer         675911140852    18.52   
...                  ..                ...                  ...      ...   
2018-12-11 07:42:50   1     Robert Johnson     3517111172421930    14.92   
2018-04-07 07:18:20  12        Megan Price     5297187379298983    14.80   
2018-08-02 07:13:49   7        Sean Taylor     3516952396080247    14.42   
2018-12-15 08:34:15  25    Nancy Contreras      372414832802279    14.36   
2018-04-24 07:41:49  15        Kyle Tucker     6500236164848279    14.30   
2018-10-22 07:41:56  13        John Martin     4711773125020499    13.83   
2018-09-03 07:27:55  12        Megan Price      376027549341849    13.59   
2018-03-10 08:52:09   2      Shane Shaffer         675911140852    13.53   
2018-04-27 08:08:38  19        Peter Mckay     3561072557118696    13.52   
2018-02-26 07:31:20  20      Kevin Spencer  4586962917519654607    13.46   
2018-10-27 08:04:44  23         Mark Lewis     4150721559116778    12.98   
2018-05-01 07:40:00  15        Kyle Tucker     6500236164848279    12.98   
2018-03-30 08:12:28   5        Sara Cooper     4268491956169254    12.96   
2018-01-17 07:31:03  18      Malik Carlson        4498002758300    12.67   
2018-08-08 08:36:23  11     Brandon Pineda      180098539019105    12.52   
2018-07-30 08:49:16   5        Sara Cooper        4276466390111    12.50   
2018-11-30 08:52:20   7        Sean Taylor     4539990688484983    12.32   
2018-12-29 08:11:55  18      Malik Carlson        4498002758300    12.25   
2018-02-19 08:50:36   1     Robert Johnson  4761049645711555811    12.17   
2018-01-20 08:35:34  12        Megan Price         501879657465    11

In [60]:
px.scatter(suspect_df, x=suspect_df.index, y='amount', color='category', size='amount', hover_data=['name', 'category'], labels = {'x' : 'Date', 'y' : 'Amount'}, title='Early Hour Transactions')

   ## Data Analysis Questions 1

   Use `hvPlot` to create a line plot showing a time series from the transactions along all the year for **card holders 2 and 18**. In order to contrast the patterns of both card holders, create a line plot containing both lines. What difference do you observe between the consumption patterns? Does the difference could be a fraudulent transaction? Explain your rationale.

In [41]:
# loading data for card holder 2 and 18 from the database
query = f'SELECT a.id, a.name, b.card, c.date, c.amount, e.name as "category" \
        FROM public.card_holder a, public.credit_card b, public.transaction c, public.merchant d, public.merchant_category e \
        WHERE a.id = b.id_card_holder AND b.card=c.card AND c.id_merchant=d.id AND d.id_merchant_category=e.id AND a.id IN (2, 18) AND b.card=c.card'

transaction_df = execute_query(query)
transaction_df.head()

data_by_id = transaction_df.groupby('id')
data_by_id

In [61]:
# plot for cardholder 2

cardholder_2 = data_by_id.get_group(2)['amount'].hvplot(ylabel='Amount', xlabel='Date', title="Transactions by Id Holder 2")
cardholder_2


:Curve   [date]   (amount)

In [62]:
# plot for cardholder 18

cardholder_18 = data_by_id.get_group(18)['amount'].hvplot(ylabel='Amount', xlabel='Date', title="Transactions by Id Holder 18")
cardholder_18

:Curve   [date]   (amount)

In [71]:
# combined plot for card holders 2 and 18
cardholder_2 * cardholder_18

:Overlay
   .Curve.Amount.I  :Curve   [date]   (amount)
   .Curve.Amount.II :Curve   [date]   (amount)

   ### Conclusions for Question 1



The consumption pattern for both the id holder is very different. Id Holder 2 makes too many small transactions. Id Holder 18 has transactions ranging till $1839. Id Holder 2 is more suspectable to fraudulent transactions

   ## Data Analysis Question 2

   Use `Plotly Express` to create a series of six box plots, one for each month, in order to identify how many outliers could be per month for **card holder id 25**. By observing the consumption patters, do you see any anomalies? Write your own conclusions about your insights.

In [67]:
# loading data of daily transactions from jan to jun 2018 for card holder 25
card_holder_id = 25
start_date = '2018-01-01' 
end_date = '2018-07-01' 
query = f'SELECT a.id, a.name, b.card, c.date, c.amount, e.name as "category" \
        FROM public.card_holder a, public.credit_card b, public.transaction c, public.merchant d, public.merchant_category e \
        WHERE a.id = b.id_card_holder AND b.card=c.card AND c.id_merchant=d.id AND d.id_merchant_category=e.id AND a.id={card_holder_id} \
        AND date BETWEEN \'{start_date} 00:00:00\'::timestamp AND \'{end_date} 00:00:00\'::timestamp'
transaction_df = execute_query(query)
transaction_df.head()

id             name             card  amount    category
date                                                                         
2018-01-02 02:06:21  25  Nancy Contreras    4319653513507    1.46  food truck
2018-01-05 06:26:45  25  Nancy Contreras  372414832802279   10.74  food truck
2018-01-07 14:57:23  25  Nancy Contreras    4319653513507    2.93  food truck
2018-01-10 00:25:40  25  Nancy Contreras  372414832802279    1.39  restaurant
2018-01-14 05:02:22  25  Nancy Contreras  372414832802279   17.84  food truck

In [68]:
# change the numeric month to month names
transaction_df['Month'] = transaction_df.index.month_name()
transaction_df.head()

id             name             card  amount    category  \
date                                                                            
2018-01-02 02:06:21  25  Nancy Contreras    4319653513507    1.46  food truck   
2018-01-05 06:26:45  25  Nancy Contreras  372414832802279   10.74  food truck   
2018-01-07 14:57:23  25  Nancy Contreras    4319653513507    2.93  food truck   
2018-01-10 00:25:40  25  Nancy Contreras  372414832802279    1.39  restaurant   
2018-01-14 05:02:22  25  Nancy Contreras  372414832802279   17.84  food truck   

                       Month  
date                          
2018-01-02 02:06:21  January  
2018-01-05 06:26:45  January  
2018-01-07 14:57:23  January  
2018-01-10 00:25:40  January  
2018-01-14 05:02:22  January

In [69]:
# creating the six box plots using plotly express
px.box(transaction_df, y='amount', hover_data=['category','card'],x='Month', color='category', title='Transactions by Id Holder 25')

   ### Conclusions for Question 2



There seems to be fraudulent transactions pertaining to Restaurant & Food Truck category where Food Truck is ranging from $1.46 to $1046